### DISCUSSION TITLE
Literal dict vs dict()

### DISCUSSION THREAD

**Discussion by aspizu** [microsoft/pyright#4879] [Q&A]
```py
JSON = (
    None
    | bool
    | int
    | float
    | str
    | list["JSON"]
    | dict[None | bool | int | float | str, "JSON"]
)
RESPONSE = dict[None | bool | int | float | str, JSON]
```

```py
def foo(a: int) -> RESPONSE:
    return {'a': a}
```
Works but,

```py
def foo(a: int) -> RESPONSE:
    return dict(a=a)
```
Does not work.

**Comment by erictraut** [comment]
Yes, that's a result of the way the `dict` constructor is defined. When you use the `**kwargs` form, as in your second example, you're using the following constructor overload:

```python
    def __init__(self: dict[str, _VT], **kwargs: _VT) -> None: ...
```

This constrains the invariant key type to be a `str`, but your `RESPONSE` type does not use `str` as a key type.

If you want to use the constructor call, you can use the alternate form `return dict({"a": a})`. This form uses this constructor overload:

```python
    def __init__(self, __map: SupportsKeysAndGetItem[_KT, _VT]) -> None: ...
```

This form allows the constraint solver to solve both `_KT` and `_VT` without constraining `_KT` to a `str`.

**Reply by aspizu** [reply]
also first assigning it to a variable then returning it doesn't work. How do i change the RESPONSE type so it allows it.

**Reply by erictraut** [reply]
If you want to use that particular form of the constructor, you would need to change `RESPONSE` to accept only `str` as its key: `RESPONSE = dict[str, JSON]`. But that's probably not what you want. Instead, I recommend switching to a different form of the `dict` constructor as I suggested above. You can either use `dict({"a": a})` or simply `{"a": a}`.

**Reply by erictraut** [reply]
Or you could create your own helper function:
```python
def make_response(**kwargs: JSON) -> RESPONSE:
    return dict(**kwargs)

def foo(a: int) -> RESPONSE:
    return make_response(a=a)
```

**Reply by aspizu** [reply]
```py
RESPONSE = dict[str, JSON] | <old thingy>
```
This doesn't work.

**Reply by erictraut** [reply]
But this does: `RESPONSE = dict[str, JSON]`. (I'm not sure what `<old thingy>` refers to.)
